fix history import and export for real user setups

bernt-matthias opened the pull request:
Adds necessary calls to the `external_chown_script` for files and directories that need to be readable / writable by the user running the job. 

Essentially ready, but I guess some beautification could be nice (e.g., maybe marge the _chown function somehow with the chown of the job class) -- but I prefer to wait first for reviews.

@mvdbeek suggested that an integration test would be nice, but this seems out of reach for me.

Below in one of the comments I collected further questions / todos.

Note that I tested on import/export on usegalaxy.eu .. and it seemed to fail as well.

bernt-matthias committed:
first steps

bernt-matthias commented:
Currently stuck at an apparently trivial problem: 

I modified the JobExportHistoryArchive to contain the FakeDatasetAssociation:

https://github.com/galaxyproject/galaxy/blob/b6d4f70298053fb09417bac78f359453d1d7ced3/lib/galaxy/model/__init__.py#L1442

which is initialised here

https://github.com/galaxyproject/galaxy/blob/b6d4f70298053fb09417bac78f359453d1d7ced3/lib/galaxy/tools/actions/history_imp_exp.py#L119

the log.error on the foillowing line outputs something like: 

```
JobExportHistoryArchive <galaxy.model.JobExportHistoryArchive(None) at 0x2b6a0ec8e850> created with fda <galaxy.model.FakeDatasetAssociation object at 0x2b6a0eb4e590>
```

Note the `None` 

If I try to use this object 

https://github.com/galaxyproject/galaxy/blob/b6d4f70298053fb09417bac78f359453d1d7ced3/lib/galaxy/tools/evaluation.py#L89

https://github.com/galaxyproject/galaxy/blob/b6d4f70298053fb09417bac78f359453d1d7ced3/lib/galaxy/jobs/__init__.py#L1954

I always get: `AttributeError: 'JobExportHistoryArchive' object has no attribute 'fda'`

Strange is also that if I print the object I get special `<galaxy.model.JobExportHistoryArchive(55) at 0x2b6a4d59ac50>`  I.e. no `None` and a seemingly different object.

I wiped out all `pyc` files and restarted Galaxy...

bernt-matthias commented:
TODOs: 

- [ ] target 19.05, 19.09 or dev 
- [ ] Check if `model.GenomeIndexToolData` also needs special care everywhere where `JobExportHistoryArchive` does  (e.g. add `FakeDataAssociation`). See here:
https://github.com/galaxyproject/galaxy/blob/b6d4f70298053fb09417bac78f359453d1d7ced3/lib/galaxy/jobs/__init__.py#L1067
- [x] Since there is now a HDA changes to the cleanup script might be necessary.
- [ ] Do both of these lines need to be called? https://github.com/galaxyproject/galaxy/blob/c2580727060583462abc76acec00577090c16da0/lib/galaxy/jobs/__init__.py#L1789
- [ ] hardcoding of "output_file" seems quite dirty (export case)
- [ ] currently neither hidden nor deleted data sets are included, do we want to change this?

mvdbeek committed:
Outline a solution for getting FakeDatasetAssociations

bernt-matthias committed:
unseccessful test to get the working dir

bernt-matthias commented:
@mvdbeek your commit seems to solve the problem ... which brings me back to this https://github.com/galaxyproject/galaxy/pull/8567 .. LOL

My latest commit (https://github.com/galaxyproject/galaxy/pull/9001/commits/a69c608e51df146a1d64d1499a3fb059698136af) is an unsuccessful test to get the jobs working dir which results in an error: 

```
sqlalchemy.exc.ArgumentError: Object <galaxy.model.Job(9504) at 0x2ba689378510> is not legal as a SQL literal value
```

I guess there is no job yet.

bernt-matthias committed:
chown tempdir before and export job

solves https://github.com/galaxyproject/galaxy/pull/8567

bernt-matthias commented:
Last commit is a better alternative for https://github.com/galaxyproject/galaxy/pull/8567.

I guess some refactoring of the call to the external chown function could be nice. I will try to implement something in util on which everything else could depend.

Which no brings me to an error during the import:

```
galaxy.jobs ERROR 2019-11-17 17:01:38,647 [p:189723,w:0,m:1] [UnivaJobRunner.work_thread-3] Unable to cleanup job 9524
Traceback (most recent call last):
  File "lib/galaxy/jobs/__init__.py", line 1790, in cleanup
    galaxy.tools.imp_exp.JobImportHistoryArchiveWrapper(self.app, self.job_id).cleanup_after_job()
  File "lib/galaxy/tools/imp_exp/__init__.py", line 53, in cleanup_after_job
    model_store.perform_import(new_history, job=job, new_history=True)
  File "lib/galaxy/model/store/__init__.py", line 168, in perform_import
    datasets_attrs = self.datasets_properties()
  File "lib/galaxy/model/store/__init__.py", line 772, in datasets_properties
    datasets_attrs = load(open(datasets_attrs_file_name))
IOError: [Errno 2] No such file or directory: u'/gpfs1/work/songalax/galaxy-dev/database/tmp/tmpoHvvkM/datasets_attrs.txt'
```

LOL. But I guess this can also be solved with chown. But I need to investigate this.

bernt-matthias committed:
add external_chown_script to MockAppConfig

mvdbeek commented:
> * target 19.05, 19.09 or dev

dev first, we can then see what parts (or perhaps all) we want to backport.

> * Check if `model.GenomeIndexToolData` also needs special care everywhere where `JobExportHistoryArchive` does  (e.g. add `FakeDataAssociation`). See here:

Keep it simple, that can be done in another PR

> * Since there is now a HDA changes to the cleanup script might be necessary.

I don't see a new HDA here ?

> * Do both of these lines need to be called? 
>     
>       
>         [galaxy/lib/galaxy/jobs/__init__.py](https://github.com/galaxyproject/galaxy/blob/c2580727060583462abc76acec00577090c16da0/lib/galaxy/jobs/__init__.py#L1789)

Seems unrelated to the PR ? Smaller, isolated changes are often better.

> * hardcoding of "output_file" seems quite dirty (export case)

You can discover the output from the tool that is associated to the job. But that might be unnecessary complexity so long as the export tools is the only tool that falls into the `special` case. Just mark it as a constant (uppercase variable) is enough IMO.  

> * [ ]  currently neither hidden nor deleted data sets are included, do we want to change this?

Perhaps in another PR.

mvdbeek commented on lib/galaxy/tools/imp_exp/__init__.py:
```diff
@@ -1,17 +1,34 @@
 ATTRS_FILENAME_HISTORY = 'history_attrs.txt'
 
 
+def _chown(path, jeha, app, user):
```
don't we have a function like that already somewhere in the codebase ? could we re-use it ?

bernt-matthias replied:
Yeah, at least here: https://github.com/galaxyproject/galaxy/blob/b91a205f89c906ce062abc6ae5bb7b5ac73ebfa8/lib/galaxy/tools/actions/upload_common.py#L320 and here https://github.com/galaxyproject/galaxy/blob/9662164ad68b39adf5a5606a7aa8e388f6a79f1e/lib/galaxy/jobs/__init__.py#L2169 these are currently bound to a class which we can't access from here as far as I see.

I was thinking about introducing a new function (doing popen, and error handling, and maybe pwent stuff) in util that could be reused. Would this be a good idea?

mvdbeek replied:
Yes, I think that would be great.

bernt-matthias replied:
OK. Function added. Need to test on my setup. 

Will introduce this function at the other places later in a separate PR.

Currently the code is missing to call the chown also for archive_source in the import case. Would be nice, but I do not have an idea how to access the path here: https://github.com/galaxyproject/galaxy/blob/c2580727060583462abc76acec00577090c16da0/lib/galaxy/jobs/__init__.py#L1789

mvdbeek replied:
Not sure about the import archive_source, that would probably be handled somewhere during upload ?

Could you also use the new function in this PR as well ? This avoids having slightly divergent functions scattered in the code, and is also a good way to figure out if anything is missing in the new function (which looks good!).

mvdbeek commented on lib/galaxy/model/__init__.py:
```diff
@@ -1422,6 +1422,22 @@ def dataset(self):
+# uses a Dataset rather than an HDA or LDA, it's necessary to set up a
+# fake dataset association that provides the needed attributes for
+# preparing a job.
+class FakeDatasetAssociation (object):
```
You can use this one in https://github.com/galaxyproject/galaxy/blob/972d641b1abec0117de812f0200eb3bad81f1822/lib/galaxy/tools/evaluation.py#L89 now to avoid duplication

bernt-matthias replied:
The one in galaxy/lib/galaxy/tools/evaluation.py should be already gone, isn't it?

mvdbeek replied:
yup, I'm just blind ... :)

bernt-matthias committed:
create a chown function to be used everywhere

mvdbeek commented on lib/galaxy/tools/imp_exp/__init__.py:
```diff
@@ -1,10 +1,13 @@
 import tempfile
 
 from galaxy import model
-from galaxy.model import store
```
I think this was done on purpose to ease refactoring, maybe keep it ?

bernt-matthias replied:
done

mvdbeek commented on lib/galaxy/tools/imp_exp/__init__.py:
```diff
@@ -40,15 +49,15 @@ def cleanup_after_job(self):
         try:
             archive_dir = jiha.archive_dir
-            model_store = store.get_import_model_store_for_directory(archive_dir, app=self.app, user=user)
+            external_chown(archive_dir, jiha.job.user.system_user_pwent(str(getpass.getuser())),
```
```suggestion
            external_chown(archive_dir, jiha.job.user.system_user_pwent(getpass.getuser()),
```
this is already a string, on python2 and python3.

mvdbeek commented on lib/galaxy/tools/imp_exp/__init__.py:
```diff
@@ -108,9 +119,12 @@ def cleanup_after_job(self):
-                log.debug('Error deleting directory containing attribute files (%s): %s' % (temp_dir, e))
+        if not jeha:
+            return
+        external_chown(jeha.temp_directory, jeha.job.user.system_user_pwent(str(getpass.getuser())),
```
```suggestion
        external_chown(jeha.temp_directory, jeha.job.user.system_user_pwent(getpass.getuser()),
```

bernt-matthias commented:
Now the new function is used everywhere. Lets see what the tests say: for which parts (jobs, upload,...) do we have tests using a real user setup? 

Note that I removed the chmod 777 in case of a failure in chown of the job dir (I think this should only happen in case of misconfigurations)

bernt-matthias commented:
Regarding archive_source. I think the last trace is here: https://github.com/galaxyproject/galaxy/blob/27c6c71214af3c1ff9e6c7c093bbc1cb3096eb4f/lib/galaxy/tools/actions/history_imp_exp.py#L54

in the following lines the incoming dict is somehow added to the job. Wondering if it can retrieved from within JobImportHistoryArchiveWrapper....

bernt-matthias commented:
Not sure how to fix the unit tests. 

Seems that in the tests for the  `jiha` which is set here:

https://github.com/galaxyproject/galaxy/blob/c6062abe5768a5460e6cc57e477121a8d127bae5/lib/galaxy/tools/imp_exp/__init__.py#L43

`jiha.job` has a None user attribute.

Also just adding a `model.User(...)`  to the job here https://github.com/galaxyproject/galaxy/blob/adaf0fe31c2afd4f8405a950c72fdea8db1e34b0/test/unit/tools/test_history_imp_exp.py#L32

does not help (odd anyway that model.Job has no user attribute).

bernt-matthias commented:
integration and selenium failures seem unrelated.

mvdbeek commented:
@bernt-matthias bumping this to 20.05, but let me know if I can do something to help. Given this is a bug I'm happy to re-target after branching. I do think this fixes history import already for `outputs_to_working_directory`, which seems like a pretty nice fix on its own.

bernt-matthias commented:
@mvdbeek agreed. Missing is a test for these cases, e.g. for  `outputs_to_working_directory` and real user setups.

mvdbeek commented:
I can create an integration test for outputs_to_working_directory, the real user setup seems a little trickier.

mvdbeek committed:
Drop JEHA.cleanup_after_job, use extra_files_path

of the created archive_dir. This means no external dangling
tempdir that needs to be chowned and deleted.
Also fixes the unit tests.

mvdbeek committed:
Split History API test into history and import/export tests

and add integration test that exercises history import/export using
outputs_to_working_directory.

bernt-matthias commented:
Hey @mvdbeek seems that this is it :) api and integration tests pass. The failing tests are unrelated, or?

bernt-matthias commented:
Would like to squash a few more commits. Would make it easier for me to cherry pick the commits to my release branch.

bernt-matthias commented on lib/galaxy/jobs/__init__.py:
```diff
@@ -2199,37 +2196,19 @@ def __link_file_check(self):
-                os.chmod(self.working_directory, RWXRWXRWX)
+        if job.user is not None:
+            external_chown(self.working_directory, self.user_system_pwent,
+                           external_chown_script, description="working directory")
```
Important question: Is it OK not to `os.chmod(self.working_directory, RWXRWXRWX)` in case of error? Or should this be integrated in `external_chown` or just here?

mvdbeek replied:
Is this change critical for fixing hisitory import/export ? If not I would keep the current behavior and change it in another PR.

bernt-matthias committed:
fix MRO problems

mvdbeek commented on lib/galaxy/tools/imp_exp/export_history.py:
```diff
@@ -26,6 +26,9 @@ def create_archive(export_directory, out_file, gzip=False):
         return 1
+    finally:
+        if os.path.exists(export_directory):
+            shutil.rmtree(export_directory, ignore_errors=True)
```
You don't need that with `ignore_errors=True`
```suggestion
        if os.path.exists(export_directory):
```
```suggestion
        shutil.rmtree(export_directory, ignore_errors=True)
```

mvdbeek commented on lib/galaxy/util/path/__init__.py:
```diff
@@ -343,6 +345,26 @@ def canonicalize(self, ext):
+    if external_chown_script is None:
+        return
+
+    try:
```
You could use commands.execute here, right? (https://github.com/galaxyproject/galaxy/pull/9544)

bernt-matthias replied:
> Is this change critical for fixing hisitory import/export ? If not I would keep the current behavior and change it in another PR.

Could be a good idea. But then `external_chown` needs a proper return value: maybe `True` for success and `False` for failed executions or `external_chown_script is None`

mvdbeek reviewed (commented):
Nice, almost there!

mvdbeek commented on lib/galaxy/tools/actions/upload_common.py:
```diff
@@ -396,8 +377,10 @@ def _chown(path):
             # user cannot remove it unless the parent directory is writable.
-            if link_data_only == 'copy_files' and trans.app.config.external_chown_script:
-                _chown(uploaded_dataset.path)
+            if link_data_only == 'copy_files':
```
```suggestion
            if link_data_only == 'copy_files' and trans.user:
```
should fix the selenium tests.

bernt-matthias committed:
Apply suggestions from code review

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

bernt-matthias committed:
use commands.execute instead of 'bare' Popen

bernt-matthias committed:
restore making file world writable in case of error

mvdbeek reviewed (approved):
Great, thanks a lot @bernt-matthias!
Let me know if this works and I'll merge it.

bernt-matthias committed:
check if archive source is URL

bernt-matthias committed:
add missing import

bernt-matthias committed:
fix rebasing errors

ApiTestCase has been moved and I missed the additional base class
which is necessary for resolving the MRO

mvdbeek commented on lib/galaxy_test/api/_framework.py:
```diff
@@ -1,3 +1,5 @@
+from unittest import TestCase
```
I think that might be not quite right ? Did this maybe go wrong with the rebase ?

bernt-matthias replied:
Actually this got lost during the rebase. If I just use `class ApiTestCase(FunctionalTestCase, UsesApiTestCaseMixin):` instead of class `ApiTestCase(FunctionalTestCase, UsesApiTestCaseMixin, TestCase):` I get 


```
____________________ ERROR collecting test/integration/test_history_import_export.py ____________________
test/integration/test_history_import_export.py:5: in <module>
    class ImportExportHistoryOutputsToWorkingDirTestCase(ImportExportHistoryTestCase, IntegrationTestCase):
E   TypeError: Cannot create a consistent method resolution
E   order (MRO) for bases TestCase, UsesApiTestCaseMixin
```

But maybe there is another solution for this Problem -- but I was unable to find one for days.

mvdbeek replied:
OK, no, I think that's fine then

mvdbeek commented on lib/galaxy/jobs/__init__.py:
```diff
@@ -1932,8 +1930,9 @@ def get_output_path(self, dataset):
-            return dataset.file_name
+#         if getattr(dataset, "fake_dataset_association", False):
+#             log.error("GET_OUTPUT_PATH FAKE %s" %(dataset.file_name))
+#             return dataset.file_name
```
```suggestion
```

mvdbeek reviewed (approved):
If you can drop the commented out code this is good to go!

bernt-matthias committed:
Apply suggestions from code review

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>
